js-envs-test-kit/src/main/scala/org/scalajs/jsenv/test/ComTests.scala

```diff
@@ -2,246 +2,147 @@ package org.scalajs.jsenv.test
 
 import org.scalajs.jsenv._
 
-import org.scalajs.core.tools.io._
-
 import org.junit.Test
 import org.junit.Assert._
 
-import scala.concurrent.Await
-
-/** A couple of tests that test communication for mix-in into a test suite */
-trait ComTests extends AsyncTests {
-
-  protected def newJSEnv: ComJSEnv
-
-  protected def comRunner(code: String): ComJSRunner = {
-    val codeVF = new MemVirtualJSFile("testScript.js").withContent(code)
-    newJSEnv.comRunner(codeVF :: Nil)
-  }
-
-  private def assertThrowClosed(msg: String, body: => Unit): Unit = {
-    val thrown = try {
-      body
-      false
-    } catch {
-      case _: ComJSEnv.ComClosedException =>
-        true
-    }
+import scala.concurrent.ExecutionContext.Implicits.global
 
-    assertTrue(msg, thrown)
-  }
+abstract class ComTests(jsEnv: JSEnv) {
+  private val kit = new TestComKit(jsEnv)
 
   @Test
-  def comCloseJVMTest: Unit = {
-    val com = comRunner(s"""
+  def basicTest: Unit = {
+    val run = kit.start("""
       scalajsCom.init(function(msg) { scalajsCom.send("received: " + msg); });
       scalajsCom.send("Hello World");
-    """)
-
-    start(com)
-
-    assertEquals("Hello World", com.receive())
+    """, RunConfig())
 
-    for (i <- 0 to 10) {
-      com.send(i.toString)
-      assertEquals(s"received: $i", com.receive())
+    try {
+      assertEquals("Hello World", run.waitNextMessage())
+
+      for (i <- 0 to 10) {
+        run.run.send(i.toString)
+        assertEquals(s"received: $i", run.waitNextMessage())
+      }
+    } finally {
+      run.closeAndWait()
     }
-
-    com.close()
-    com.await(DefaultTimeout)
-
-    com.stop() // should do nothing, and not fail
-  }
-
-  def comCloseJSTestCommon(timeout: Long): Unit = {
-    val com = comRunner(s"""
-      scalajsCom.init(function(msg) {});
-      for (var i = 0; i < 10; ++i)
-        scalajsCom.send("msg: " + i);
-      scalajsCom.close();
-    """)
-
-    start(com)
-
-    Thread.sleep(timeout)
-
-    for (i <- 0 until 10)
-      assertEquals(s"msg: $i", com.receive())
-
-    assertThrowClosed("Expect receive to throw after closing of channel",
-        com.receive())
-
-    com.close()
-    com.await(DefaultTimeout)
-
-    com.stop() // should do nothing, and not fail
   }
 
-  @Test
-  def comCloseJSTest: Unit = comCloseJSTestCommon(0)
-
-  @Test
-  def comCloseJSTestDelayed: Unit = comCloseJSTestCommon(1000)
-
-  @Test
-  def doubleCloseTest: Unit = {
-    val n = 10
-    val com = pingPongRunner(n)
-
-    start(com)
-
-    for (i <- 0 until n) {
-      com.send("ping")
-      assertEquals("pong", com.receive())
-    }
-
-    com.close()
-    com.await(DefaultTimeout)
-  }
+  // TODO: What should happen if the JS side terminates?
```

Comment:
@sjrd Any advice on this? Also, we need a way to inject VM-side termination into the test, since this is not specified in the `JSEnv` interface (and a VM does not need to support it).

Reply:
Note that this is notably causing the CI to fail right now in `library/test`.

Reply:
You can define in `ComTests` a method like:
```scala
def getTerminateVMJSCode(): Option[String] = None
```
which can be overridden by subclasses of `ComTests` to provide the appropriate termination code. By default, `None` indicates that is not supported. The test can then call that method, and `assume` it `isDefined`.

Now about what should be the nice behavior, well ... I guess we should complete the `run.future` with a successful value. Any issue with doing that?